src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,299 @@
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using NuGet.Common;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol.Model;
+
+namespace NuGet.Protocol.Resources
+{
+    /// <summary>Implementation of <see cref="IVulnerabilityInfoResource"/> for NuGet V3 HTTP feeds.</summary>
+    /// <remarks>Not intended to be created directly. Use <see cref="SourceRepository.GetResourceAsync{T}(CancellationToken)"/>
+    /// with <see cref="IVulnerabilityInfoResource"/> for T, and typecast to this class.
+    /// <para>Implements the <a href="https://learn.microsoft.com/nuget/api/vulnerability-info-resource">VulnerabilityInfo server API resource</a>.</para></remarks>
+    public sealed class VulnerabilityInfoResourceV3 : IVulnerabilityInfoResource
+    {
+        private readonly SourceRepository _sourceRepository;
+
+        internal VulnerabilityInfoResourceV3(SourceRepository sourceRepository)
+        {
+            _sourceRepository = sourceRepository ?? throw new ArgumentNullException(nameof(sourceRepository));
+        }
+
+        /// <summary>Get the vulnerability pages the server contains.</summary>
+        /// <param name="cacheContext">The cache settings to use when making HTTP requests.</param>
+        /// <param name="log">The logger for any messages.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <returns>The list of vulnerability data pages the server claims to have.</returns>
+        /// <exception cref="FatalProtocolException">When various HTTP or deserialization exceptions occur.</exception>
+        public async Task<IReadOnlyList<V3VulnerabilityIndexEntry>> GetVulnerabilityFilesAsync(SourceCacheContext cacheContext, ILogger log, CancellationToken cancellationToken)
+        {
+            Uri vulnerabilityIndexUrl = await GetIndexUrlAsync(cancellationToken);
+            HttpSourceResource httpSourceResource = await _sourceRepository.GetResourceAsync<HttpSourceResource>(cancellationToken);
+
+            HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, isFirstAttempt: true);
+            var request = new HttpSourceCachedRequest(vulnerabilityIndexUrl.OriginalString, "vuln_index", httpSourceCacheContext);
+            IReadOnlyList<V3VulnerabilityIndexEntry>? vulnFiles;
+            try
+            {
+                vulnFiles = await httpSourceResource.HttpSource.GetAsync(request,
+                    result =>
+                {
+                    using (var textReader = new StreamReader(result.Stream))
+                    using (var jsonReader = new JsonTextReader(textReader))
+                    {
+                        var parsed = JsonExtensions.JsonObjectSerializer.Deserialize<IReadOnlyList<V3VulnerabilityIndexEntry>>(jsonReader);
+                        return Task.FromResult(parsed);
+                    }
+                },
```

Comment:
Needs further indentation since this code block is related to the result parameter and not vulnFiles.
```suggestion
                    {
                        using (var textReader = new StreamReader(result.Stream))
                        using (var jsonReader = new JsonTextReader(textReader))
                        {
                            var parsed = JsonExtensions.JsonObjectSerializer.Deserialize<IReadOnlyList<V3VulnerabilityIndexEntry>>(jsonReader);
                            return Task.FromResult(parsed);
                        }
                    },
```